File: src/main/java/gg/mineral/practice/util/items/ItemBuilder.java
package gg.mineral.practice.util.items;

import java.util.Arrays;
import java.util.List;

import org.bukkit.Material;
import org.bukkit.inventory.ItemStack;

import lombok.val;

public class ItemBuilder {
    private Material material;
    private int durability, amount;
    private String name;
    private List<String> lore = null;

    public ItemBuilder(ItemStack item) {
        this.material = item.getType();
        this.durability = item.getDurability();
        this.amount = item.getAmount();
    }

    public ItemBuilder(Material material) {
        this(new ItemStack(material));
    }

    public static ItemBuilder from(ItemStack item) {
        return new ItemBuilder(item);
    }

    public ItemBuilder amount(int amount) {
        this.amount = amount;
        return this;
    }

    public ItemBuilder durability(int durability) {
        this.durability = durability;
        return this;
    }

    public ItemBuilder name(String name) {
        this.name = name;
        return this;
    }

    public ItemBuilder lore(String... lore) {
        this.lore = Arrays.asList(lore);
        return this;
    }

    public ItemStack build() {
        val newItemStack = new ItemStack(material, amount, (short) durability);
        val meta = newItemStack.getItemMeta();
        meta.setDisplayName(name);
        meta.setLore(lore);
        newItemStack.setItemMeta(meta);
        return newItemStack;
    }
}

File: src/main/kotlin/gg/mineral/practice/util/items/ItemBuilder.kt
package gg.mineral.practice.util.items

import org.bukkit.Material
import org.bukkit.inventory.ItemStack

class ItemBuilder(item: ItemStack) {
    private val material: Material = item.type
    private var durability: Int
    private var amount: Int
    private var name: String? = null
    private var lore: List<String>? = null

    init {
        this.durability = item.durability.toInt()
        this.amount = item.amount
    }

    constructor(material: Material) : this(ItemStack(material))

    fun amount(amount: Int): ItemBuilder {
        this.amount = amount
        return this
    }

    fun durability(durability: Int): ItemBuilder {
        this.durability = durability
        return this
    }

    fun name(name: String?): ItemBuilder {
        this.name = name
        return this
    }

    fun lore(vararg lore: String): ItemBuilder {
        this.lore = listOf(*lore)
        return this
    }

    fun build(): ItemStack {
        val newItemStack = ItemStack(material, amount, durability.toShort())
        val meta = newItemStack.itemMeta
        meta.displayName = name
        meta.lore = lore
        newItemStack.setItemMeta(meta)
        return newItemStack
    }

    companion object {
        fun from(item: ItemStack): ItemBuilder {
            return ItemBuilder(item)
        }
    }
}
